src/winter_onboarding_2021/fleet_management_service/handlers/user.clj

```diff
@@ -0,0 +1,83 @@
+  (:require [ring.util.response :as response]
+            [clojure.spec.alpha :as s]
+            [crypto.password.bcrypt :as password]
+            [winter-onboarding-2021.fleet-management-service.session :as session]
+            [winter-onboarding-2021.fleet-management-service.views.user :as view]
+            [winter-onboarding-2021.fleet-management-service.models.user :as user-model]
+            [winter-onboarding-2021.fleet-management-service.specs :as specs]))
+
+(defn flash-msg [msg success?]
+  (if success?
+    {:flash {:success true
+             :style-class "alert alert-success"
+             :message msg}}
+    {:flash {:error true
+             :style-class "alert alert-danger"
+             :message msg}}))
+
+(defn signup-form [_]
+  {:title "Sign-up"
+   :content (view/signup-form)})
+
+(defn user-exist? [email]
+  (not-empty (user-model/find-by-keys {:email email})))
+
+(defn uuid []
+  (java.util.UUID/randomUUID))
+
+(defn create-user [{:keys [multipart-params]}]
+  (let [validated-user (s/conform ::specs/signup-form multipart-params)
+        user-exist (user-exist? (:email multipart-params))]
+    (if (or (s/invalid? validated-user) user-exist)
+      (if user-exist
+        (-> (flash-msg "User already exisits, use different email!" false)
+            (merge (response/redirect "/users/signup")))
+        (-> (flash-msg "Could not create user, enter valid details!" false)
+            (merge (response/redirect "/users/signup"))))
+      (let [_ (user-model/create (assoc validated-user
+                                        :role "admin"
+                                        :password (password/encrypt
+                                                   (:password validated-user))))]
+        (merge (flash-msg "User created successfully!" true) (response/redirect "/users/signup"))))))
+
+(defn login-form [_]
+  {:title "Login"
+   :content (view/login-form)})
+
+(defn- successful-login-response [user-id]
+  (let [session-id (uuid)]
+    (session/write-session session-id user-id)
+    (merge (flash-msg "Hooray! Logged in!" true)
+           (-> (response/redirect "/users/dashboard")
+               (response/set-cookie "session-id" session-id {:path "/"})))))
+
+(defn- wrong-password-response []
+  (merge (flash-msg "Wrong password" false)
+         (response/redirect "/users/login")))
+
+(defn- email-not-found-response []
+  (merge (flash-msg "User with email not found" false)
+         (response/redirect "/users/login")))
+
+(defn- invalid-data-response []
+  (merge (flash-msg "Please send valid data" false)
+         (response/redirect "/users/login")))
+
+(defn login [{:keys [params]}]
+  (let [validated-params (s/conform ::specs/login-params params)]
+    (if (s/invalid? validated-params)
+      (invalid-data-response)
+      (if-let [db-user (first (user-model/find-by-keys (select-keys validated-params [:email])))]
+        (if (password/check (:password validated-params) (:users/password db-user))
```

Comment:
So if we move this logic to a model, it will return something like this:
`{:found? true :user db-user} ` for when password is correct

`{:found? true :user nil}`  for the case when password is wrong

`{:found? false :user nil}` for the case when email is not found.

The handler will get the above structures from model & it will then return the appropriate response.

cc: @yogirajh007